File: trunk/cpp/src/sfntly/table/bitmap/index_sub_table.cc
/*
 * Copyright 2011 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "sfntly/table/bitmap/index_sub_table.h"

#include "sfntly/table/bitmap/eblc_table.h"
#include "sfntly/table/bitmap/index_sub_table_format1.h"
#include "sfntly/table/bitmap/index_sub_table_format2.h"
#include "sfntly/table/bitmap/index_sub_table_format3.h"
#include "sfntly/table/bitmap/index_sub_table_format4.h"
#include "sfntly/table/bitmap/index_sub_table_format5.h"

namespace sfntly {

bool IndexSubTable::CheckGlyphRange(int32_t glyph_id) {
  if (glyph_id < first_glyph_index() || glyph_id > last_glyph_index()) {
#if !defined (SFNTLY_NO_EXCEPTION)
    throw IndexOutOfBoundException("Glyph ID is outside of the allowed range");
#endif
    return false;
  }
  return true;
}

template <typename IndexSubTableType>
static IndexSubTableType* CreateTable(ReadableFontData* data,
                                      int32_t index_sub_table_offset,
                                      int32_t first_glyph_index,
                                      int32_t last_glyph_index) {
  ReadableFontDataPtr new_data;
  new_data.Attach(down_cast<ReadableFontData*>(
      data->Slice(index_sub_table_offset,
                  IndexSubTableType::GetDataLength(data,
                                                   index_sub_table_offset,
                                                   first_glyph_index,
                                                   last_glyph_index))));
  return new IndexSubTableType(new_data, first_glyph_index, last_glyph_index);
}

// static
CALLER_ATTACH IndexSubTable*
    IndexSubTable::CreateIndexSubTable(ReadableFontData* data,
                                       int32_t offset_to_index_sub_table_array,
                                       int32_t array_index) {
  int32_t index_sub_table_entry_offset =
      offset_to_index_sub_table_array +
      array_index * EblcTable::Offset::kIndexSubTableEntryLength;

  int32_t first_glyph_index =
      data->ReadUShort(index_sub_table_entry_offset +
                       EblcTable::Offset::kIndexSubTableEntry_firstGlyphIndex);
  int32_t last_glyph_index =
      data->ReadUShort(index_sub_table_entry_offset +
                       EblcTable::Offset::kIndexSubTableEntry_lastGlyphIndex);
  int32_t additional_offset_to_index_subtable = data->ReadULongAsInt(
      index_sub_table_entry_offset +
      EblcTable::Offset::kIndexSubTableEntry_additionalOffsetToIndexSubTable);

  int32_t index_sub_table_offset = offset_to_index_sub_table_array +
                                   additional_offset_to_index_subtable;

  int32_t index_format = data->ReadUShort(index_sub_table_offset);
  IndexSubTablePtr new_table;
  ReadableFontDataPtr new_data;
  switch (index_format) {
    case 1:
      new_table = CreateTable<IndexSubTableFormat1>(data,
                                                    index_sub_table_offset,
                                                    first_glyph_index,
                                                    last_glyph_index);
      break;

    case 2:
      new_table = CreateTable<IndexSubTableFormat2>(data,
                                                    index_sub_table_offset,
                                                    first_glyph_index,
                                                    last_glyph_index);
      break;

    case 3:
      new_table = CreateTable<IndexSubTableFormat3>(data,
                                                    index_sub_table_offset,
                                                    first_glyph_index,
                                                    last_glyph_index);
      break;

    case 4:
      new_table = CreateTable<IndexSubTableFormat4>(data,
                                                    index_sub_table_offset,
                                                    first_glyph_index,
                                                    last_glyph_index);
      break;

    case 5:
      new_table = CreateTable<IndexSubTableFormat5>(data,
                                                    index_sub_table_offset,
                                                    first_glyph_index,
                                                    last_glyph_index);
      break;

    default:
      // Unknown format and unable to process.
#if !defined (SFNTLY_NO_EXCEPTION)
      throw IllegalArgumentException("Invalid Index Subtable Format");
#endif
      break;
  }

  return new_table.Detach();
}

IndexSubTable::IndexSubTable(ReadableFontData* data,
                             int32_t first,
                             int32_t last)
    : SubTable(data), first_glyph_index_(first), last_glyph_index_(last) {
  index_format_ =
      data_->ReadUShort(EblcTable::Offset::kIndexSubHeader_indexFormat);
  image_format_ =
      data_->ReadUShort(EblcTable::Offset::kIndexSubHeader_imageFormat);
  image_data_offset_ =
      data_->ReadULongAsInt(EblcTable::Offset::kIndexSubHeader_imageDataOffset);
}

}  // namespace sfntly

File: java/src/com/google/typography/font/sfntly/table/bitmap/IndexSubTable.java
/*
 * Copyright 2011 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.typography.font.sfntly.table.bitmap;

import com.google.typography.font.sfntly.data.ReadableFontData;
import com.google.typography.font.sfntly.data.WritableFontData;
import com.google.typography.font.sfntly.table.SubTable;
import java.util.Iterator;

public abstract class IndexSubTable extends SubTable {
  private static final boolean DEBUG = false;

  public interface Format {
    int FORMAT_1 = 1;
    int FORMAT_2 = 2;
    int FORMAT_3 = 3;
    int FORMAT_4 = 4;
    int FORMAT_5 = 5;
  }

  private interface SubOffset {
    int indexFormat = 0;
    int imageFormat = 2;
    int imageDataOffset = 4;
  }

  private final int firstGlyphIndex;
  private final int lastGlyphIndex;
  private final int indexFormat;
  private final int imageFormat;
  private final int imageDataOffset;

  protected static IndexSubTable createIndexSubTable(
      ReadableFontData data, int offsetToIndexSubTableArray, int arrayIndex) {

    Builder<? extends IndexSubTable> builder =
        Builder.createBuilder(data, offsetToIndexSubTableArray, arrayIndex);
    if (builder == null) {
      return null;
    }
    return builder.build();
  }

  protected IndexSubTable(ReadableFontData data, int firstGlyphIndex, int lastGlyphIndex) {
    super(data);
    this.firstGlyphIndex = firstGlyphIndex;
    this.lastGlyphIndex = lastGlyphIndex;
    this.indexFormat = this.data.readUShort(SubOffset.indexFormat);
    this.imageFormat = this.data.readUShort(SubOffset.imageFormat);
    this.imageDataOffset = this.data.readULongAsInt(SubOffset.imageDataOffset);
  }

  public int indexFormat() {
    return indexFormat;
  }

  public int firstGlyphIndex() {
    return firstGlyphIndex;
  }

  public int lastGlyphIndex() {
    return lastGlyphIndex;
  }

  public int imageFormat() {
    return imageFormat;
  }

  public int imageDataOffset() {
    return imageDataOffset;
  }

  public BitmapGlyphInfo glyphInfo(int glyphId) {
    int loca = checkGlyphRange(glyphId);
    if (loca == -1) {
      return null;
    }
    if (glyphStartOffset(glyphId) == -1) {
      return null;
    }

    return new BitmapGlyphInfo(
        glyphId, imageDataOffset(), glyphStartOffset(glyphId), glyphLength(glyphId), imageFormat());
  }

  public final int glyphOffset(int glyphId) {
    int glyphStartOffset = glyphStartOffset(glyphId);
    if (glyphStartOffset == -1) {
      return -1;
    }
    return imageDataOffset() + glyphStartOffset;
  }

  /** Gets the offset of the glyph relative to the block for this index subtable. */
  public abstract int glyphStartOffset(int glyphId);

  public abstract int glyphLength(int glyphId);

  public abstract int numGlyphs();

  protected static int checkGlyphRange(int glyphId, int firstGlyphId, int lastGlyphId) {
    if (glyphId < firstGlyphId || glyphId > lastGlyphId) {
      throw new IndexOutOfBoundsException("Glyph ID is outside of the allowed range.");
    }
    return glyphId - firstGlyphId;
  }

  protected int checkGlyphRange(int glyphId) {
    return checkGlyphRange(glyphId, firstGlyphIndex(), lastGlyphIndex());
  }

  @Override
  public String toString() {
    String s =
        String.format(
            "IndexSubTable: [%#x : %#x], format = %d, image format = %s, imageOff = %#x%n",
            firstGlyphIndex(), lastGlyphIndex(), indexFormat, imageFormat(), imageDataOffset());

    if (DEBUG) {
      for (int g = firstGlyphIndex(); g < lastGlyphIndex(); g++) {
        s += String.format("\tgid = %d, offset = %d%n", g, glyphStartOffset(g));
      }
    }
    return s;
  }

  public abstract static class Builder<T extends IndexSubTable> extends SubTable.Builder<T> {
    private int firstGlyphIndex;
    private int lastGlyphIndex;
    private int indexFormat;
    private int imageFormat;
    private int imageDataOffset;

    public static Builder<? extends IndexSubTable> createBuilder(int indexFormat) {
      switch (indexFormat) {
        case 1:
          return IndexSubTableFormat1.Builder.createBuilder();
        case 2:
          return IndexSubTableFormat2.Builder.createBuilder();
        case 3:
          return IndexSubTableFormat3.Builder.createBuilder();
        case 4:
          return IndexSubTableFormat4.Builder.createBuilder();
        case 5:
          return IndexSubTableFormat5.Builder.createBuilder();
        default:
          throw new IllegalArgumentException(
              String.format("Invalid index subtable format %d", indexFormat));
      }
    }

    static Builder<? extends IndexSubTable> createBuilder(
        ReadableFontData data, int offsetToIndexSubTableArray, int arrayIndex) {

      int entryOffset = offsetToIndexSubTableArray + arrayIndex * EblcTable.IndexSubTableEntry.SIZE;

      int firstGlyphIndex =
          data.readUShort(entryOffset + EblcTable.IndexSubTableEntry.firstGlyphIndex);
      int lastGlyphIndex =
          data.readUShort(entryOffset + EblcTable.IndexSubTableEntry.lastGlyphIndex);
      int additionOffsetToIndexSubtable =
          data.readULongAsInt(
              entryOffset + EblcTable.IndexSubTableEntry.additionalOffsetToIndexSubtable);

      int indexSubTableOffset = offsetToIndexSubTableArray + additionOffsetToIndexSubtable;

      int indexFormat = data.readUShort(indexSubTableOffset);
      switch (indexFormat) {
        case 1:
          return IndexSubTableFormat1.Builder.createBuilder(
              data, indexSubTableOffset, firstGlyphIndex, lastGlyphIndex);
        case 2:
          return IndexSubTableFormat2.Builder.createBuilder(
              data, indexSubTableOffset, firstGlyphIndex, lastGlyphIndex);
        case 3:
          return IndexSubTableFormat3.Builder.createBuilder(
              data, indexSubTableOffset, firstGlyphIndex, lastGlyphIndex);
        case 4:
          return IndexSubTableFormat4.Builder.createBuilder(
              data, indexSubTableOffset, firstGlyphIndex, lastGlyphIndex);
        case 5:
          return IndexSubTableFormat5.Builder.createBuilder(
              data, indexSubTableOffset, firstGlyphIndex, lastGlyphIndex);
        default:
          throw new IllegalArgumentException(
              String.format("Invalid index subtable format %d", indexFormat));
      }
    }

    protected Builder(int dataSize, int indexFormat) {
      super(dataSize);
      this.indexFormat = indexFormat;
    }

    protected Builder(int indexFormat, int imageFormat, int imageDataOffset, int dataSize) {
      this(dataSize, indexFormat);
      this.imageFormat = imageFormat;
      this.imageDataOffset = imageDataOffset;
    }

    protected Builder(WritableFontData data, int firstGlyphIndex, int lastGlyphIndex) {
      super(data);
      this.firstGlyphIndex = firstGlyphIndex;
      this.lastGlyphIndex = lastGlyphIndex;
      initialize(data);
    }

    protected Builder(ReadableFontData data, int firstGlyphIndex, int lastGlyphIndex) {
      super(data);
      this.firstGlyphIndex = firstGlyphIndex;
      this.lastGlyphIndex = lastGlyphIndex;
      initialize(data);
    }

    private void initialize(ReadableFontData data) {
      this.indexFormat = data.readUShort(SubOffset.indexFormat);
      this.imageFormat = data.readUShort(SubOffset.imageFormat);
      this.imageDataOffset = data.readULongAsInt(SubOffset.imageDataOffset);
    }

    /**
     * Unable to fully revert unless some changes happen to hold the original data. Until then keep
     * as protected.
     */
    protected void revert() {
      setModelChanged(false);
      initialize(internalReadData());
    }

    public int indexFormat() {
      return indexFormat;
    }

    public int firstGlyphIndex() {
      return firstGlyphIndex;
    }

    public void setFirstGlyphIndex(int firstGlyphIndex) {
      this.firstGlyphIndex = firstGlyphIndex;
    }

    public int lastGlyphIndex() {
      return lastGlyphIndex;
    }

    public void setLastGlyphIndex(int lastGlyphIndex) {
      this.lastGlyphIndex = lastGlyphIndex;
    }

    public int imageFormat() {
      return imageFormat;
    }

    public void setImageFormat(int imageFormat) {
      this.imageFormat = imageFormat;
    }

    public int imageDataOffset() {
      return imageDataOffset;
    }

    public void setImageDataOffset(int offset) {
      this.imageDataOffset = offset;
    }

    public abstract int numGlyphs();

    /** Gets the glyph info for the specified glyph id. */
    // TODO(stuartg): could be optimized by pushing down into subclasses
    public BitmapGlyphInfo glyphInfo(int glyphId) {
      return new BitmapGlyphInfo(
          glyphId,
          imageDataOffset(),
          glyphStartOffset(glyphId),
          glyphLength(glyphId),
          imageFormat());
    }

    /** Gets the full offset of the glyph within the EBDT table. */
    public final int glyphOffset(int glyphId) {
      return imageDataOffset() + glyphStartOffset(glyphId);
    }

    /** Gets the offset of the glyph relative to the block for this index subtable. */
    public abstract int glyphStartOffset(int glyphId);

    /** Gets the length of the glyph within the EBDT table. */
    public abstract int glyphLength(int glyphId);

    /**
     * Checks that the glyph id is within the correct range. If it returns the offset of the glyph
     * id from the start of the range.
     *
     * @return the offset of the glyphId from the start of the glyph range
     * @throws IndexOutOfBoundsException if the glyph id is not within the correct range
     */
    protected int checkGlyphRange(int glyphId) {
      return IndexSubTable.checkGlyphRange(glyphId, firstGlyphIndex(), lastGlyphIndex());
    }

    protected int serializeIndexSubHeader(WritableFontData data) {
      int size = data.writeUShort(SubOffset.indexFormat, indexFormat);
      size += data.writeUShort(SubOffset.imageFormat, imageFormat);
      size += data.writeULong(SubOffset.imageDataOffset, imageDataOffset);
      return size;
    }

    abstract Iterator<BitmapGlyphInfo> iterator();

    /*
     * The following methods will never be called but they need to be here to
     * allow the BitmapSizeTable to see these methods through an abstract
     * reference.
     */
    @Override
    protected T subBuildTable(ReadableFontData data) {
      return null;
    }

    @Override
    protected void subDataSet() {
      // NOP
    }

    @Override
    protected int subDataSizeToSerialize() {
      return 0;
    }

    @Override
    protected boolean subReadyToSerialize() {
      return false;
    }

    @Override
    protected int subSerialize(WritableFontData newData) {
      return 0;
    }

    @Override
    public String toString() {
      String s =
          String.format(
              "IndexSubTable: [%#x : %#x], format = %d, image format = %s, imageOff = %#x%n",
              firstGlyphIndex(), lastGlyphIndex(), indexFormat, imageFormat(), imageDataOffset());

      if (DEBUG) {
        for (int g = firstGlyphIndex(); g < lastGlyphIndex(); g++) {
          s += String.format("\tgid = %d, offset = %d%n", g, glyphStartOffset(g));
        }
      }
      return s;
    }
  }
}
